### PR TITLE
Add ability to build within a container

### PR DESCRIPTION
**Pull request by jdekonin**
Signed-off-by: Joe deKoning <joe_dekoning@ca.ibm.com>

### CODE REVIEW

**Comment by jdekonin** [comment]
jenkins compile xlinux jdk8

File: buildenv/jenkins/common/build.groovy [hunk 1]
```diff
@@ -585,26 +585,75 @@ def cleanWorkspace (KEEP_WORKSPACE) {
...
+            sh "docker pull ${DOCKER_IMAGE}"
+        }
+    }
+    DOCKER_IMAGE_ID = get_docker_image_id(DOCKER_IMAGE)
+    
+    if (!DOCKER_IMAGE_ID) {
+        println("List docker images")
+        sh "docker images"
+        error("Invalid docker image: ${DOCKER_IMAGE}!")
+    } else {
```
Comment (vsebe) [inline buildenv/jenkins/common/build.groovy:617]: The `else` clause is not required due to error() function in the if branch; just echo... outside if statement.

File: buildenv/jenkins/common/build.groovy [hunk 2]
```diff
@@ -585,26 +585,75 @@ def cleanWorkspace (KEEP_WORKSPACE) {
...
+
+def prepare_docker_environment() {
+    clean_docker_containers()
+    DOCKER_IMAGE_ID = get_docker_image_id(DOCKER_IMAGE)
+
+    // if there is no id found then attempt to pull it
+    if (!DOCKER_IMAGE_ID) {
+        echo "${DOCKER_IMAGE} not found locally, attempting to pull from dockerhub"
+        docker.withRegistry("", "9f50c848-8764-440d-b95a-1d295c21713e") {
+            sh "docker pull ${DOCKER_IMAGE}"
```
Comment (vsebe) [inline buildenv/jenkins/common/build.groovy:608]: Wrap it in `withCredentials` and add for docker artifactory.

**Review by vsebe — CHANGES_REQUESTED** [review]

File: buildenv/jenkins/common/build.groovy [hunk 3]
```diff
@@ -585,26 +585,75 @@ def cleanWorkspace (KEEP_WORKSPACE) {
...
+    println("Listing docker containers to attempt removal")
+    sh "docker ps -a"
+    def containers = sh (script: "docker ps -a --format \"{{.ID}}\"", returnStdout: true).trim()
+    if (containers) {
+        println("Stop all docker containers")
+        sh "docker ps -a --format \"{{.ID}}\" | xargs docker stop"
+        println("Remove all docker containers")
+        sh "docker ps -a --format \"{{.ID}}\" | xargs docker rm --force"
+    }
+}
```
Comment (AdamBrousseau) [inline buildenv/jenkins/common/build.groovy:598]: How does this differ from what we've used in the past for other jobs?
```
docker system prune -af
```
Are we trying to cleanup everything or just containers/images for this job?

File: buildenv/jenkins/common/build.groovy (hunk 2 above)
Comment (AdamBrousseau) [inline buildenv/jenkins/common/build.groovy:608]: Should we pull everytime? If a container is updated I believe it won't update on a machine unless we pull explicitly.

File: buildenv/jenkins/common/build.groovy [hunk 4]
```diff
@@ -585,26 +585,75 @@ def cleanWorkspace (KEEP_WORKSPACE) {
...
+        docker.withRegistry("", "9f50c848-8764-440d-b95a-1d295c21713e") {
+            sh "docker pull ${DOCKER_IMAGE}"
+        }
+    }
+    DOCKER_IMAGE_ID = get_docker_image_id(DOCKER_IMAGE)
+    
+    if (!DOCKER_IMAGE_ID) {
+        println("List docker images")
+        sh "docker images"
+        error("Invalid docker image: ${DOCKER_IMAGE}!")
```
Comment (AdamBrousseau) [inline buildenv/jenkins/common/build.groovy:616]: Will this error ever be hit? Won't the pull error out if it fails or doesn't exist?

File: buildenv/jenkins/common/variables-functions.groovy [hunk 5]
```diff
@@ -550,7 +551,13 @@ def set_node(job_type) {
         if (!NODE) {
             error("Missing ${job_type} NODE!")
         }
+    }
 
+    if (NODE.contains('docker')) {
```
Comment (AdamBrousseau) [inline buildenv/jenkins/common/variables-functions.groovy:556]: I'm not sure why this `if`. Shouldn't we only care if the variable file has `docker_image` for the spec? I think this means the node name must contain `docker`, which is not always true.

File: buildenv/jenkins/variables/defaults.yml [hunk 6]
```diff
@@ -395,12 +395,16 @@ x86-64_linux:
...
+      11: 'ci.role.build && hw.arch.x86 && sw.tool.docker'
+    docker_image:
+      all: 'eclipseopenj9/jenkins-agent-x86-centos6:13'
   extra_configure_options:
     all: '--enable-jitserver'
   build_env:
-    cmd: 'source /home/jenkins/set_gcc7.5.0_env'
+    cmd:
+      16: 'source /home/jenkins/set_gcc7.5.0_env'
+      next: 'source /home/jenkins/set_gcc7.5.0_env'
```
Comment (AdamBrousseau) [inline buildenv/jenkins/variables/defaults.yml:407]: Why this change?

File: buildenv/jenkins/common/build.groovy (hunk 2 above)
Comment (jdekonin) [reply buildenv/jenkins/common/build.groovy:608]: Updated in my next push

File: buildenv/jenkins/common/build.groovy (hunk 3 above)
Comment (jdekonin) [reply buildenv/jenkins/common/build.groovy:598]: This just covers stopping and removing any running containers, which is really not that necessary as the job does that on its own if it completes cleanly.  I'm not sure we want to add a `system prune` to each job as that would delete and force a fresh download of an image each time when its rarely going to be required.

File: buildenv/jenkins/common/build.groovy (hunk 2 above)
Comment (jdekonin) [reply buildenv/jenkins/common/build.groovy:608]: The version number is specified in the defaults.yaml, in this case `jenkins-agent-x86-centos6:13`.  The pull is only required when the host system does not have that image local.

File: buildenv/jenkins/common/build.groovy (hunk 4 above)
Comment (jdekonin) [reply buildenv/jenkins/common/build.groovy:616]: I can remove, you are correct we shouldn't hit this and if we did the docker login/pull would indicate a failure.  Updated in my next push.

File: buildenv/jenkins/variables/defaults.yml (hunk 6 above)
Comment (jdekonin) [reply buildenv/jenkins/variables/defaults.yml:407]: The resources that the `set_gcc7.5.0_env` scripts setups up on static hosts is the default in the docker container.

**Comment by jdekonin** [comment]
jenkins compile xlinux jdk8

**Comment by jdekonin** [comment]
jenkins compile xlinux jdk8

File: buildenv/jenkins/common/variables-functions.groovy [hunk 7]
```diff
@@ -547,10 +547,10 @@ def set_node(job_type) {
     if (!NODE) {
         // fetch from variables file
         NODE = buildspec_manager.getSpec(SPEC).getScalarField("node_labels.${job_type}", SDK_VERSION)
+        DOCKER_IMAGE = buildspec_manager.getSpec(SPEC).getScalarField("node_labels.docker_image", SDK_VERSION) ?: ''
```
[the remark this answers is not in the document]
Comment (jdekonin) [reply buildenv/jenkins/common/variables-functions.groovy:550]: I guess it depends on the usage we want.  For manual use in debugging we often set the Node to a specific machine which then this code isn't executed.  For a real build this is executed as per the passing PR job.  I've moved the line outside the !NODE check which get the behaviour I think we need for both use cases.

File: buildenv/jenkins/common/variables-functions.groovy (hunk 5 above)
Comment (jdekonin) [reply buildenv/jenkins/common/variables-functions.groovy]: Reworked and removed.

**Comment by jdekonin** [comment]
jenkins compile xlinux,zlinux jdk8,jdk11,jdk16

**Comment by jdekonin** [comment]
jenkins test sanity xlinux,zlinux jdk8,jdk11,jdk16

**Comment by jdekonin** [comment]
The JDK16 failed testing doesn't appear to be related at all.  I believe this is ready.

**Review by AdamBrousseau — APPROVED** [review]

File: buildenv/jenkins/variables/defaults.yml [hunk 8]
```diff
@@ -395,12 +396,17 @@ x86-64_linux:
   node_labels:
     build:
       all: 'ci.role.build && hw.arch.x86 && sw.os.cent.7'
-      8: 'ci.role.build && hw.arch.x86 && sw.os.cent.6'
-      11: 'ci.role.build && hw.arch.x86 && sw.os.cent.6'
+      8: 'ci.role.build && hw.arch.x86 && sw.tool.docker'
+      11: 'ci.role.build && hw.arch.x86 && sw.tool.docker'
+    docker_image:
+      8: 'eclipseopenj9/jenkins-agent-x86-centos6:13'
+      11: 'eclipseopenj9/jenkins-agent-x86-centos6:13'
```
Comment (AdamBrousseau) [inline buildenv/jenkins/variables/defaults.yml:403]: tag 13 or 14?

File: buildenv/jenkins/common/variables-functions.groovy (hunk 7 above)
[the remark this answers is not in the document]
Comment (jdekonin) [reply buildenv/jenkins/common/variables-functions.groovy]: I also changed the docker_image value per release in defaults.yaml to avoid this issue.

File: buildenv/jenkins/variables/defaults.yml (hunk 8 above)
Comment (jdekonin) [reply buildenv/jenkins/variables/defaults.yml]: Should be 14 now since that one is built based on head of stream repo content and not my PR. :)  Updated.

**Comment by jdekonin** [comment]
jenkins compile xlinux,zlinux jdk8,jdk11,jdk16

**Review by vsebe — APPROVED** [review]

**Comment by jdekonin** [comment]
jenkins compile xlinux,zlinux jdk8,jdk11,jdk16

**Comment by jdekonin** [comment]
@pshipton ready for review/merge.

**Review by pshipton — APPROVED** [review]
